frontend/src/metabase/query_builder/components/filters/pickers/DefaultPicker.jsx

```diff
@@ -61,6 +64,42 @@ export default function DefaultPicker({
   const isBetweenLayout =
     operator.name === "between" && operatorFields.length === 2;
 
+  const extractFieldSettings = field => {
+    const visualizationSettings = filter
+      ?.query()
+      ?.question()
+      ?.settings();
+
+    // map column settings by field id
+    const parsedVisualizationSettings = Object.fromEntries(
+      Object.entries(visualizationSettings?.column_settings || {}).map(
+        ([key, value]) => {
+          try {
+            // destringify the keys eg: '["ref",["field",39,null]]' to find field ids
+            const fieldId = JSON.parse(key)[1][1];
+            return [fieldId, value];
+          } catch (err) {
+            return [null, null];
+          }
+        },
+      ),
+    );
```

Comment:
I couldn't find any function the decodes these mbql keys :disappointed: , the encoding function is in `/lib/settings/visualization.js`

Reply:
There is a function `keyForColumn` that can get a key to receive the column settings. It also handles other types of `field_refs` and I would use it:

```js
  const visualizationSettings = filter
    ?.query()
    ?.question()
    ?.settings();
  const columnSettings =
    visualizationSettings.column_settings[
      keyForColumn(filter.field().column())
    ];
  const fieldMetadata = field?.metadata?.fields[field?.id];
  const fieldSettings = {
    ...(fieldMetadata?.settings ?? {}),
    ...(columnSettings ?? {}),
  };

  const currencyPrefix =
    isCurrency(field) && fieldSettings?.currency
      ? getCurrencySymbol(fieldSettings.currency)
      : null;
```

Reply:
Unfortunately, that function proves to be a bit less versatile than extracting the field ids from decoded keys: namely, it doesn't properly pick up visualization settings for joined fields.

![Screenshot from 2022-04-05 13-53-40](https://user-images.githubusercontent.com/30528226/161838635-0e1b0a8b-7028-414b-b3ee-f0ce360f420d.png)

It looks like `keyForColumn()` naively encodes the field ref into mbql using only the field ref or name instead of extracting the field ids from the stringified keys :cry: